### PR TITLE
treewide: Replace uses of `replace-literal` with `replace-secret` to avoid leaking secrets

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
Using `replace-literal` (or similar) to insert secrets leaks the secrets through the `replace-literal` process' `/proc/<pid>/cmdline` file. This introduces a tiny `replace-secret` utility which solves this by reading the secret straight from the file instead, which also usually simplifies the code a bit, since we're always reading the secret from a file anyway.

I've focused on replacing uses of `replace-literal` here, since they translate very easily, but `sed` usage and other similar ones should also be replaced in the future.

The `mpd` and `mpdscribble` changes have not been tested with real credentials, so help from someone who uses them would be appreciated.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/replace-secret/replace-secret.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+@click.argument('secret_filename')
+@click.argument('filename')
+def replace(string_to_replace, secret_filename, filename):
+    """Replace a string in one file with a secret from a second file.
+
+    Since the secret is read from a file, it won't be leaked through
+    '/proc/<pid>/cmdline', unlike when 'sed' or 'replace' is used.
+
+    """
+    with open(secret_filename) as sf, open(filename, 'r+') as f:
```
Comment (grahamc) [inline pkgs/build-support/replace-secret/replace-secret.py:16]: Should this set a umask or something?

Comment (talyz) [reply pkgs/build-support/replace-secret/replace-secret.py:16]: It does the replacement in-place, so the file's permissions retained. I guess it could force the permissions to something restrictive, but this feels like a surprising behavior..

**Review by doronbehar — APPROVED** [review]
Diff LGTM.

File: pkgs/build-support/replace-secret/replace-secret.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{ stdenv, lib, python3 }:
+
+stdenv.mkDerivation {
+  name = "replace-secret";
+  buildInputs = [ (python3.withPackages (p: [ p.click ]))];
```
Comment (doronbehar) [inline pkgs/build-support/replace-secret/replace-secret.nix:5]: Is it possible to make this package not depend on `click` perhaps?

Comment (talyz) [reply pkgs/build-support/replace-secret/replace-secret.nix]: Sure, I switched to `argparse` instead.

File: pkgs/build-support/replace-secret/replace-secret.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        old = f.read()
+        secret = sf.read()
+        while secret[-1] == '\n':
```
[the remark this answers is not in the document]
Comment (talyz) [reply pkgs/build-support/replace-secret/replace-secret.py]: Yeah, that works. :) It strips from the front too, but that's probably not an issue.

**Comment by fricklerhandwerk** [comment]
LGTM

File: pkgs/build-support/replace-secret/replace-secret.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{ stdenv, lib, python3 }:
+
+stdenv.mkDerivation {
+  name = "replace-secret";
+  buildInputs = [ python3 ];
+  buildCommand = ''
+    mkdir -p $out/bin
+    cp ${./replace-secret.py} $out/bin/replace-secret
```
Comment (doronbehar) [inline pkgs/build-support/replace-secret/replace-secret.nix:8]:
```suggestion
    install -D ${./replace-secret.py} $out/bin/replace-secret
```

File: nixos/modules/services/web-apps/discourse.nix [hunk 5]
```diff
@@ -713,11 +710,12 @@ in
                   cfg.siteSettings
                   "/run/discourse/config/nixos_site_settings.json"
               }
-              install -T -m 0400 -o discourse ${discourseConf} /run/discourse/config/discourse.conf
+              install -T -m 0600 -o discourse ${discourseConf} /run/discourse/config/discourse.conf
```
Comment (doronbehar) [inline nixos/modules/services/web-apps/discourse.nix:713]: Is this change related?

File: nixos/modules/services/web-apps/discourse.nix [hunk 6]
```diff
@@ -713,11 +710,12 @@ in
...
               ${mkSecretReplacement cfg.database.passwordFile}
               ${mkSecretReplacement cfg.mail.outgoing.passwordFile}
               ${mkSecretReplacement cfg.redis.passwordFile}
               ${mkSecretReplacement cfg.secretKeyBaseFile}
+              chmod 0400 /run/discourse/config/discourse.conf
```
Comment (doronbehar) [inline nixos/modules/services/web-apps/discourse.nix:718]: This change too perhaps is worth merging, but should at least come in a separate commit.

**Review by doronbehar — COMMENTED** [review]
All commits touching `replace-secret.py` should be squashed.

File: nixos/modules/services/web-apps/discourse.nix (hunk 5 above)
Comment (talyz) [reply nixos/modules/services/web-apps/discourse.nix:713]: Yes, this is required for it to be able to write to the file. `replace-literal` ignores permissions (or maybe the `-f` flag also has that effect), but `replace-secret` doesn't.

File: nixos/modules/services/web-apps/discourse.nix (hunk 6 above)
Comment (talyz) [reply nixos/modules/services/web-apps/discourse.nix:718]: This makes sure the file has the same permissions as before the commit, but sets it after inserting the secrets.

**Review by doronbehar — APPROVED** [review]

**Review by fricklerhandwerk — APPROVED** [review]

File: pkgs/build-support/replace-secret/replace-secret.nix [hunk 7]
```diff
@@ -0,0 +1,22 @@
+{ stdenv, lib, python3 }:
```
Comment (andir) [inline pkgs/build-support/replace-secret/replace-secret.nix:1]: Perhaps add a check phase here that does some trivial usage of the command so we can be sure it still works without having to run nixos tests.

**Review by andir — CHANGES_REQUESTED** [review]

File: pkgs/build-support/replace-secret/replace-secret.nix (hunk 7 above)
Comment (talyz) [reply pkgs/build-support/replace-secret/replace-secret.nix:1]: Yeah, that's a good idea. I added two simple checks. If the diff looks okay to you, I'll squash it.
